lib/backend/k8s/resources/workloadendpoint.go

```diff
@@ -165,59 +170,108 @@ func (c *WorkloadEndpointClient) Get(ctx context.Context, key model.Key, revisio
 	if !c.converter.IsValidCalicoWorkloadEndpoint(pod) {
 		return nil, cerrors.ErrorResourceDoesNotExist{Identifier: k}
 	}
-	return c.converter.PodToWorkloadEndpoint(pod)
+
+	kvps, err := c.converter.PodToWorkloadEndpoints(pod)
+	if err != nil {
+		return nil, err
+	}
+
+	// Find the WorkloadEndpoint that has a name matching the name in the given key
+	for _, kvp := range kvps {
+		wep := kvp.Value.(*apiv3.WorkloadEndpoint)
+		if wep.Name == key.(model.ResourceKey).Name {
+			return kvp, nil
+		}
+	}
+
+	return nil, kerrors.NewNotFound(apiv3.Resource("WorkloadEndpoint"), key.String())
 }
 
 func (c *WorkloadEndpointClient) List(ctx context.Context, list model.ListInterface, revision string) (*model.KVPairList, error) {
 	log.Debug("Received List request on WorkloadEndpoint type")
 	l := list.(model.ResourceListOptions)
+	var podName string
 
-	// If a workload is provided, we can do an exact lookup of this
-	// workload endpoint.
+	// If a "Name" is provided, we may be able to get the exact WorkloadEndpoint
 	if l.Name != "" {
-		kvp, err := c.Get(ctx, model.ResourceKey{
-			Name:      l.Name,
-			Namespace: l.Namespace,
-			Kind:      l.Kind,
-		}, revision)
+		wepID, err := c.converter.ParseWorkloadEndpointName(l.Name)
 		if err != nil {
-			switch err.(type) {
-			// Return empty slice of KVPair if the object doesn't exist, return the error otherwise.
-			case cerrors.ErrorResourceDoesNotExist:
+			return nil, err
+		}
+
+		// Pods can have multiple WorkloadEndpoints, so check that the "Endpoint" name is specified, otherwise we'll look
```

Comment:
I'm not sure if this is correct - if a `Name` and `Namespace` are both provided, that is an _exact_ identifer for a workload endpoint. We can't do any more specific lookup.

This is apparent by the fact that the `model.ResourceKey` that you pass in to the Get query only has those keys.

Reply:
The name provided might not be the exact name, it might only be a prefix of the name (which excludes the interface name), and that prefix could uniquely identify a pod, and since the pods and workload used to have a one to one relationship that was good enough to uniquely identify a WorkloadEndpoint.

If you look at how Get worked before you'll see that we don't use the Endpoint. A good example of this scenerio and is in the CNI plugin.

Reply:
I completely agree with restructuring, I'll do that.

> Otherwise, if the endpoint is not present, and neither is the Pod, we return an empty list - I think this case is not correct. We should return an error, OR return all WEPs on the provided node

 Yes you're right, this case is wrong. I was initially thinking this is what would happen before the these changes, but I see that's not correct, not sure what I was seeing when I did this before.

Reply:
Just thinking about this again, the list options could specify this as prefixed when it's really a uniquely identifying name, so I don't think the logic should be split on if it's a prefix or not, we need to know if the Endpoint name was found regardless.

Reply:
added "list" and "listUsingName" functions to split up the logic, let me know what you think.